src/number_theory/legendre_symbol/norm_num.lean

```diff
@@ -0,0 +1,460 @@
+      (zc, nc, er, p) ← prove_jacobi_sym_odd zc nc amb eb, -- compute `jacobi_sym_nat (a % b) b`
+      pure (zc, nc, er, `(jacobi_sym_nat.mod_left).mk_app [ea, eb, amb, er, phab, p])
+    else
+    prove_jacobi_sym_odd zc nc ea eb
+  | _ := failed
+  end
+
+/-- This evaluates `r := jacobi_sym a b` and produces a proof term for the equality.
+This is done by reducing to `r := jacobi_sym_nat (a % b) b`. -/
+meta def prove_jacobi_sym : instance_cache → instance_cache → expr → expr
+    → tactic (instance_cache × instance_cache × expr × expr)
+| zc nc ea eb := do
+  match match_numeral eb with -- deal with simple cases right away
+  | match_numeral_result.zero := pure (zc, nc, `(1 : ℤ), `(jacobi_sym.zero_right).mk_app [ea])
+  | match_numeral_result.one := pure (zc, nc, `(1 : ℤ), `(jacobi_sym.one_right).mk_app [ea])
+  | _ := do -- Now `1 < b`. Compute `jacobi_sym_nat (a % b) b` instead.
+    b ← eb.to_nat,
+    (zc, eb') ← zc.of_int (b : ℤ),
+    -- Get the proof that `(b : ℤ) = b'` (where `eb'` is the numeral representing `b'`).
+    -- This is important to avoid inefficient matching between the two.
+    (zc, nc, eb₁, pb') ← prove_nat_uncast zc nc eb',
+    (zc, amb, phab) ← prove_div_mod zc ea eb' tt, -- compute `a % b`
+    (zc, nc, amb', phab') ← prove_nat_uncast zc nc amb, -- `a % b` as a natural number
+    (zc, nc, er, p) ← prove_jacobi_sym_nat zc nc amb' eb₁, -- compute `jacobi_sym_nat (a % b) b`
+    pure (zc, nc, er,
+          `(jacobi_sym.mod_left).mk_app [ea, eb₁, amb', amb, er, eb', pb', phab, phab', p])
+  end
+
+end norm_num
+
+end evaluation
+
+section tactic
+
+/-!
+### The `norm_num` plug-in
+-/
+
+namespace tactic
+namespace norm_num
+
+/-- This is the `norm_num` plug-in that evaluates Jacobi and Legendre symbols. -/
+@[norm_num] meta def eval_jacobi_sym : expr → tactic (expr × expr)
+| `(jacobi_sym %%ea %%eb) := do -- Jacobi symbol
+    zc ← mk_instance_cache `(ℤ),
+    nc ← mk_instance_cache `(ℕ),
+    (prod.snd ∘ prod.snd) <$> norm_num.prove_jacobi_sym zc nc ea eb
+| `(norm_num.jacobi_sym_nat %%ea %%eb) := do -- Jacobi symbol on natural numbers
+    zc ← mk_instance_cache `(ℤ),
+    nc ← mk_instance_cache `(ℕ),
+    (prod.snd ∘ prod.snd) <$> norm_num.prove_jacobi_sym_nat zc nc ea eb
+| `(@legendre_sym %%ep %%inst %%ea) := do -- Legendre symbol
+    zc ← mk_instance_cache `(ℤ),
+    nc ← mk_instance_cache `(ℕ),
+    (zc, nc, er, pf) ← norm_num.prove_jacobi_sym zc nc ea ep,
+    pure (er, `(norm_num.legendre_sym.to_jacobi_sym).mk_app [ep, inst, ea, er, pf])
+| _ := failed
+
+end norm_num
+end tactic
+
+end tactic
+
+section examples
+
+open_locale number_theory_symbols
+
+example : J(123 | 335) = -1 := by norm_num
+example : J(-2345 | 6789) = -1 := by norm_num
+example : J(-1 | 1655801) = 1 := by norm_num
+example : J(-102334155 | 165580141) = -1 := by norm_num
+
+example : J(58378362899022564339483801989973056405585914719065 |
+            53974350278769849773003214636618718468638750007307) = -1 := by norm_num
```

Comment:
This section should go in the `test` directory

Reply:
OK; it seems that `test/norm_num_ext.lean` should be the right place for that.
I'll add the import and the examples there.